.github/workflows/openstack/prune_openstack

```diff
@@ -1,84 +1,93 @@
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack keypair delete "$ID"
-        fi
-    done < <(cat keyid_created_name.txt)
-}
+# Static constants
+use constant {
+    OS_AUTH_TYPE            => 'v3applicationcredential',
+    OS_AUTH_URL             => 'https://keystone.hou-01.cloud.prod.cpanel.net:5000/v3',
+    OS_IDENTITY_API_VERSION => '3',
+    OS_REGION_NAME          => 'RegionOne',
+    OS_INTERFACE            => 'public',
+    KEY_NAME                => "deletethis",
+    VM_NAME                 => "elevate.github.cpanel.net",
+};
 
-get_volumes () {
-    rm volume_ids.txt > /dev/null 2>&1
-    openstack volume list | grep packer  | grep available | awk '{print $2}' >> volume_ids.txt
-}
+my $openstack_path = "/usr/bin/openstack" if -x "/usr/bin/openstack";
+die "No Openstack CLI binary installed." unless $openstack_path;
+
+# Define "two hours ago" timestamp
+my $time_offset = 7200;
+my $hammer_time = POSIX::strftime( "%Y-%m-%dT%H:%M:%SZ", gmtime( time - $time_offset ) );
+
+# Run removal functions
+remove_stale_entities( 'keys',      KEY_NAME, 'keypair' );
+remove_stale_entities( 'instances', VM_NAME,  'server' );
+
+# Fetch and format entities
+sub get_entities {
+    my ($entity_type) = @_;
+
+    my $entities = {};
 
-remove_volumes () {
-    while IFS= read -r ID; do
-        openstack volume delete "$ID" || true
-    done < <(cat volume_ids.txt)
+    my ( $list_command, $show_command );
+    if ( $entity_type eq 'keys' ) {
+        $list_command = "$openstack_path keypair list -f json";
+        $show_command = "$openstack_path keypair show -f json";
+    }
+    else {
+        $list_command = "$openstack_path server list -f json --no-name-lookup";
+        $show_command = "$openstack_path server show -f json";
+    }
+
+    my $out  = _run_cmd( $list_command, 0 );
+    my $json = JSON::PP->new->decode($out);
+
+    foreach my $entry (@$json) {
+        my $entity_name = length $entry->{ID} ? $entry->{ID} : $entry->{Name};
+        my $show_out    = _run_cmd( qq{$show_command "$entity_name"}, 0 );
+        my $entity_data = JSON::PP->new->decode($show_out);
+
+        $entities->{$entity_name} = {
+            created_on => $entity_data->{created_at},
+            name       => $entity_data->{name},
+        };
+        if ( length $entity_data->{id} ) {
+            $entities->{$entity_name}->{id} = $entity_data->{id};
+        }
+    }
+    return $entities;
 }
 
-cleanup () {
-    rm -fv id_created_name.txt
-    rm -fv keyid_created_name.txt
+# Remove stale entities
+sub remove_stale_entities {
+    my ( $entity_type, $name_pattern, $delete_command ) = @_;
+
+    my $entities = get_entities($entity_type);
+
+    foreach my $entity ( grep { $_->{name} =~ /${\$name_pattern}/ } values %{$entities} ) {
+        if ( $entity->{created_on} lt $hammer_time ) {
+            print "## [INFO]: Deleting: ID: $entity->{id}, Name: $entity->{name}, Created On: $entity->{created_on}\n";
```

Comment:
Need to guard against `$entity->{id}` being undef

Reply:
It will always be defined.  For keys ID and Name are identical and for VMs the ID will be unique whereas the Name will not be.

Reply:
`$VAR1 = \{
            'created_on' => '2024-11-19T23:08:55.000000',
            'id' => 'Ee9MurByb0bMmA35KI9Hmy-deletethis',
            'name' => 'Ee9MurByb0bMmA35KI9Hmy-deletethis'
          };
`

`
$VAR1 = \{
            'name' => 'elevate.github.cpanel.net',
            'created_on' => '2024-11-19T23:08:57Z',
            'id' => 'fdf28948-be4c-4071-bc05-ca6cac61a4b5'
          };
          `

Reply:
Guarding against it anyway.

Reply:
Adding this guard helped uncover a bug:

```
$ perl prune_openstack

[ REDACTED ]

## [INFO]: Running command: /usr/bin/openstack server show -f json "6000730f-7aed-45ef-ba6b-253dba96c402"
## [WARN]: Skipping entity due to missing keys
## [WARN]: Skipping entity due to missing keys
## [WARN]: Skipping entity due to missing keys
## [WARN]: Skipping entity due to missing keys
## [WARN]: Skipping entity due to missing keys
## [WARN]: Skipping entity due to missing keys
```

Dumping out the data revealed that `created_on` is undef:

```
$VAR1 = {
          'name' => 'elevate.github.cpanel.net',
          'created_on' => undef,
          'id' => '0c100b40-3238-498e-9cce-175159a55612'
        };
```

This only happens on the server calls:

* ssh key is good
```
# openstack keypair show -f json "just_another_ssh_key"
{
  "created_at": "2021-05-21T19:42:12.000000",
  "fingerprint": "70:d7:af:af:bf:b4:1b:f5:95:ea:ba:73:bb:8d:b3:68",
  "id": "id_rsa",
  "is_deleted": false,
  "name": "id_rsa",
  "private_key": null,
  "type": "ssh",
  "user_id": "ae7599de4cd280a9ba0f191cee0511db87db0f44da1b95a2a8b7033a4e86ad45"
}
```

* server uses the `created` key instead of `created_at`
```
# openstack server show -f json "6000730f-7aed-45ef-ba6b-253dba96c402" | grep created
  "created": "2021-05-21T15:42:56Z",
```

LIkely need to do something similar to the following:
```
--- a/.github/workflows/openstack/prune_openstack
+++ b/.github/workflows/openstack/prune_openstack
@@ -56,7 +56,7 @@ sub get_entities {
         my $entity_data = JSON::PP->new->decode($show_out);

         $entities->{$entity_name} = {
-            created_on => $entity_data->{created_at},
+            created_on => $entity_data->{created_at} ? $entity_data->{created_at} : $entity_data->{created},
             name       => $entity_data->{name},
         };
         if ( length $entity_data->{id} ) {
```

Reply:
Man that is an EXCELLENT catch.  I wonder how long I've leaned on these openstack CLI clients and they were missing such data.